File: MPI/sieve/sieve_parallel_Pipeline.c
#include <stdio.h>
#include "mpi.h"

#define N 12
#define ROOT 0
#define TERMINATE -1
#define TAG 100

void display_time(double start, double end);
 
int main(int argc, char *argv[]){
    int rank, size;
    MPI_Status status;
    double begin, end;

    MPI_Init(&argc, &argv);
	MPI_Comm_size(MPI_COMM_WORLD, &size);
	MPI_Comm_rank(MPI_COMM_WORLD, &rank);

    if (rank == ROOT){
        begin = MPI_Wtime();
        int i, x=2;
        (void) printf("%d ", x);    /* Print the first prime */
        for (i=3; i<=size; ++i)
            if (i % x != 0)
                MPI_Send(&i, 1, MPI_INT, rank+1, TAG, MPI_COMM_WORLD);  /* Send primes to P1 */

        i = TERMINATE;
        MPI_Send(&i, 1, MPI_INT, rank+1, TAG, MPI_COMM_WORLD);  /* Send termination signal */
    }

    else {
        int x, i=0;
        MPI_Recv(&x, 1, MPI_INT, rank-1, TAG, MPI_COMM_WORLD, &status);   /* Receive a prime number and print it */
        (void) printf("%d ", x);
        while(i != TERMINATE){ 
            MPI_Recv(&i, 1, MPI_INT, rank-1, TAG, MPI_COMM_WORLD, &status); /* Keep receiving numbers until termination signal is received */
            if (i != TERMINATE && !(i % x)) /* If termination signal not received & i is prime */
                if (rank != size-1)
                    MPI_Send(&i, 1, MPI_INT, rank+1, TAG, MPI_COMM_WORLD);  /* Send it if it is not the last process */
                else
                    (void) printf("%d ", i);    /* Print it if it is the last process */
        }
    }

    if (rank == ROOT){
        end = MPI_Wtime();
        display_time(begin, end);   /* runtime */
    }
    
    MPI_Finalize();

    return 0;
}

/*
 * Function:  display_time 
 * --------------------
 * Prints the time (seconds) elapsed
 *
 *  start: the time (seconds) in which the whole process started
 *  end: the time (seconds) in which the whole process started
 *
 */

void display_time(double start, double end){
    (void) printf("Time spent for sorting: %g seconds\n\n", (double)(end-start));
}

File: MPI/mpi4py/sieve/sieve_parallel_Pipeline.py

import numpy as np
from mpi4py import MPI

TERMINATE = -1
TAG = 200
ROOT = 0
N = 21

if __name__ == '__main__':

    comm = MPI.COMM_WORLD
    size = comm.Get_size()
    rank = comm.Get_rank()
    status = MPI.Status()

    if rank == ROOT:
        begin = MPI.Wtime()
        x=2
        print(f'Primes numbers from 1 to {N} are : ')
        print(x) # Print the first prime 
        for i in range(x+1, N+1):
            if i % x:
                comm.send(obj=i, dest=rank+1, tag=TAG)  # Send primes to P1
        i = TERMINATE
        comm.send(obj=i, dest=rank+1, tag=TAG)  # Send termination signal
        i = comm.recv(source=size-1, tag=TAG, status=status)
        end = MPI.Wtime()

        print(f'\nTime spent for operation: {end-begin} seconds')

    if rank == (size-1):
        x=0
        while True:
            i = comm.recv(source=rank-1, tag=TAG, status=status)
            if not x: 
                x = i # Keep receiving numbers until termination signal is received
            if i == TERMINATE: # If termination signal received OR i is prime
                comm.send(obj=i, dest=ROOT, tag=TAG)
                break
            print(i) 

    if rank != ROOT and rank != (size-1):
        x=0
        while True:
            i = comm.recv(source=rank-1, tag=TAG, status=status)
            if not x: 
                x = i
                print(i)
            if i % x or i == TERMINATE: # If termination signal received OR i is prime
            	comm.send(obj=i, dest=rank+1, tag=TAG)
            if i == TERMINATE:
                break
        
        
